### PR TITLE
Warn when MaxDirectMemorySize may be incorrect (Windows/JDK8 only issue)

### LINKED ISSUE #47384: JvmErgonomicsTests.testExtractValidHeapSizeNoOptionPresent fails on 7.4
It fails with:

```
org.elasticsearch.tools.launchers.JvmErgonomicsTests > testExtractValidHeapSizeNoOptionPresent FAILED
    java.lang.AssertionError: 
    Expected: a value greater than <0L>
         but: <0L> was equal to <0L>
        at __randomizedtesting.SeedInfo.seed([22C99F519C4644E0:535C63ED1A8DEF9E]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.tools.launchers.JvmErgonomicsTests.testExtractValidHeapSizeNoOptionPresent(JvmErgonomicsTests.java:59)
```

What's strange is that this did not print any reproduction line, so I wasn't able to try the exact test.

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.4+multijob-windows-compatibility/os=windows-2019/64/consoleFull
https://gradle-enterprise.elastic.co/s/mm3dqwj56d4aa/console-log

### PR DESCRIPTION
**Pull request by williamrandolph**
There is a bug in JDK 8 for Windows where, with flags such as `-XX:+PrintFlagsFinal`, memory sizes are printed as unsigned 32-bit integers. The maximum value of such an integer is `2^32 - 1`, which is one byte less than the value of `4g` in a JVM argument. If you pass `-XX:+PrintFlagsFinal -Xmx=4g`, you will see a heap size of zero. This bug only seems to affect display; the JVM still starts with the correct heap size. On Windows and Linux, the size value is cast to an unsigned 64-bit integer, which has plenty of room for holding realistic memory sizes. In JDK 9 and beyond, [this bug was fixed][bugfix], so any memory size will be displayed correctly.

  [bugfix]: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8074459

This issue affects our JvmErgonomics code. In order to set MaxDirectMemorySize, we parse the output from `java -XX:+PrintFlagsFinal -version` (adding whichever flags we'd like to see), retrieve the value for MaxHeapSize, and set MaxDirectMemorySize to half of heap. However, if the user has provided a value for MaxDirectMemorySize, we use the provided value.

Consider a system with 40G of physical RAM. By default, the JVM would set the MaxHeapSize at one quarter of that, which is to say, 10G. Our intention would be to set MaxDirectMemorySize to 5G. But on Windows, JDK 8 will print MaxHeapSize as 2G, and MaxDirectMemorySize will be set to 1G.

The workaround is simple: set `-XX:MaxDirectMemorySize` directly.

If we wanted to fix this problem under the hood, we'd have to do two things when running on Windows with JDK8:

1. Check whether the user has passed in one of the maximum heap size flags (not just -Xmx but also -XX:MaxHeapSize). If so, parse the value for that flag and see if it's over 4g. If so, use the parsed value to set MaxDirectMemorySize.
2. If MaxHeapSize is unspecified, use an `OperatingSystemMXBean` to find the system's physical memory. Use that to set a correct heap size and a correct MaxDirectMemorySize.

Not only does the above seem like brittle logic, it also puts us in the business of duplicating Java's logic for parsing memory values. The virtue of our current implementation is that we don't have to do that at all.

Note that starting with 8.0.0, we will require a Java 11 runtime, so this problem won't happen on the 8.x line.

In light of that level of complexity, I think it's sufficient to emit a warning when running on Windows with Java 8 when there's no explicit setting for MaxDirectMemorySize. I am not sure how bad it is to have an incorrect setting (a setting of 0 in the worst case), but I could easily be persuaded that we should fail to start without an explicit MaxDirectMemorySize setting in a case where our JVM Ergonomics would set it incorrectly, but that would be a breaking change to startup configuration for some users and we'd probably have to flag it in release notes or something.

Fixes #47384

### CODE REVIEW

**Comment by williamrandolph** [comment]
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/bwc

**Comment by atorok** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: distribution/tools/launchers/src/test/java/org/elasticsearch/tools/launchers/JvmErgonomicsTests.java [hunk 1]
```diff
@@ -56,8 +56,7 @@ public void testExtractValidHeapSizeUsingMaxHeapSize() throws InterruptedExcepti
     }
 
     public void testExtractValidHeapSizeNoOptionPresent() throws InterruptedException, IOException {
-        // Muting on Windows, awaitsfix: https://github.com/elastic/elasticsearch/issues/47384
-        assumeFalse(System.getProperty("os.name").startsWith("Windows"));
+        assumeFalse(System.getProperty("os.name").startsWith("Windows") && JavaVersion.majorVersion(JavaVersion.CURRENT) == 8);
```
Comment (atorok) [inline distribution/tools/launchers/src/test/java/org/elasticsearch/tools/launchers/JvmErgonomicsTests.java]: Might be worth keeping the comment to point at the issue

File: distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmErgonomics.java [hunk 2]
```diff
@@ -58,6 +60,10 @@ private JvmErgonomics() {
         final long heapSize = extractHeapSize(finalJvmOptions);
         final long maxDirectMemorySize = extractMaxDirectMemorySize(finalJvmOptions);
         if (maxDirectMemorySize == 0) {
+            if (System.getProperty("os.name").startsWith("Windows") && JavaVersion.majorVersion(JavaVersion.CURRENT) == 8) {
+                Launchers.errPrintln("Warning: with JDK 8 on Windows, Elasticsearch may miscalculate MaxDirectMemorySize");
+                Launchers.errPrintln("  Please use a newer version of Java or set MaxDirectMemorySize explicitly");
```
Comment (atorok) [inline distribution/tools/launchers/src/main/java/org/elasticsearch/tools/launchers/JvmErgonomics.java]: Maybe link to the JDK bug or the GH issue for more information ? 
The message does not suggest this is due to the JDK

**Comment by rjernst** [comment]
Thanks for investigating this issue! A warning seems fine to me, and the code here looks good, but I'd like @danielmitterdorfer to weigh in since he has been the primary implementor of the jvm ergonomics.

**Review by danielmitterdorfer — APPROVED** [review]
Thanks for spotting this! Looks fine to me. I have just one small suggestion. We could link to the OpenJDK bug tracker instead (https://bugs.openjdk.java.net/browse/JDK-8074459) or don't include the link at all (I think we've tried to avoid this so far) and only reference the bug number `JDK-8074459`.

**Comment by williamrandolph** [comment]
@elasticmachine run elasticsearch-ci/default-distro
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Pull request merged**